trl/trainer/sft_trainer.py

```diff
@@ -0,0 +1,310 @@
+        optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
+        preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
+        peft_config: Optional[Dict] = None,
+        dataset_text_field: Optional[str] = None,
+        packing: Optional[bool] = False,
+        formatting_func: Optional[Callable] = None,
+        max_seq_length: Optional[int] = None,
+        infinite: Optional[bool] = False,
+        num_of_sequences: Optional[int] = 1024,
+        chars_per_token: Optional[float] = 3.6,
+        prepare_in_int8_kwargs: Optional[Dict] = {},
+        **pretrained_kwargs,
+    ):
+        if isinstance(model, str):
+            warnings.warn(
+                "You passed a model_id to the SFTTrainer. This will automatically create an "
+                "`AutoModelForCausalLM` or a `PeftModel` (if you passed a `peft_config`) for you."
+            )
+        elif len(pretrained_kwargs) > 1:
+            raise ValueError(
+                "You can't pass `.from_pretrained` keyword arguments to the SFTTrainer if you pass a model object.",
+                " make sure to properly initialize the model object and pass it to the SFTTrainer without additional arguments.",
+            )
+
+        if is_peft_available() and peft_config is not None:
+            if not isinstance(peft_config, PeftConfig):
+                raise ValueError(
+                    "If you want to use the PeftModel, you need to pass a PeftConfig object to the SFTTrainer."
+                    f" and you passed a {type(peft_config)}."
+                )
+
+            if not isinstance(model, PeftModel):
+                if pretrained_kwargs.get("load_in_8bit", False) and "device_map" not in pretrained_kwargs:
+                    device_map = {"": Accelerator().process_index}
+                    pretrained_kwargs["device_map"] = device_map
+
+                model = AutoModelForCausalLM.from_pretrained(
+                    model,
+                    **pretrained_kwargs,
+                )
+
+                if getattr(model, "is_loaded_in_8bit", False):
+                    model = prepare_model_for_int8_training(model, **prepare_in_int8_kwargs)
+
+                model = get_peft_model(model, peft_config)
+
+            if callbacks is None:
+                callbacks = [PeftSavingCallback]
+        elif not isinstance(model, PreTrainedModel):
+            model = AutoModelForCausalLM.from_pretrained(model, **pretrained_kwargs)
+
+        if tokenizer is None:
+            tokenizer = AutoTokenizer.from_pretrained(model.config._name_or_path)
+            if getattr(tokenizer, "pad_token", None) is None:
+                tokenizer.pad_token = tokenizer.eos_token
+
+        if max_seq_length is None:
+            # to overcome some issues with broken tokenizers
+            max_seq_length = min(tokenizer.model_max_length, 2048)
```

Comment:
haha why 2048? some models have more context window. also then we should probably emit a warning if we are not using the tokenizer's maximum.

Reply:
ahaha

Reply:
Fixed the maximum to 496 and added a warning message in [710c998](https://github.com/lvwerra/trl/pull/323/commits/710c99879860e5d02afb8dea9ba50b386f6959d4)